Implement float operators in C / Support Float and Double

Dudeldu opened the pull request:
This fixes multiple issues:
 - it reimplements ocaml round -> therefore we can get ride of the external dependency
 - it exposes c's `atof` to ocaml to make string parsing possible
 - all the operations are possible for floats and doubles and a uniform interface is provided for both
 - it would now also be fairly simple to "out-source" other functions to c like `sqrt`, `log`, `cos`, `sin` etc.
 - we now support at least the two float kinds Float and Double
 - adjusts the casting in forward evaluation as well as backwards (invariant) and also fixes some rounding issues in invariant generally
 - introduces yet another module layer to store the ftype information
 - extends the regression and unit tests to also test for floats

Builds on #9

Dudeldu committed:
Support FFloat and FDouble

- Implement float operators in C
- Support FFloat and FDouble

FelixKrayer commented on src/cdomains/floatOps/floatOps.ml:
```diff
@@ -0,0 +1,61 @@
+  let is_finite = Float.is_finite
+  let zero = 0.
+  let max_value = Float.max_float
+  let min_value = -. Float.max_float
```
I would be careful with `min_value`. Usually, when one talks about  "min value" for floats, they are talking about the least representable number greater than zero (see `DBL_MIN` for C or `Float.min_float` for OCaml). 
Also whenever someone actually needs the smallest (negative) value for a floating-point number they can just `-. max_value` since in contrast to integers the smallest and largest value have the same absolute value

Dudeldu replied:
I see your point and, personally, i dislike this naming in general (`smallest_float` would be a way better name for what is now called `min_float`). And for integer values the name `min_int` actually denotes the "highest" negative value. Also, getting the respective negative extreme value is not that easy (due to some ocaml type things...) and would look like this `Float_t.sub Nearest Float_t.zero Float_t.max_value`. I don't know, what is the best in this situation? Do you prefer only having a `max_value` and computing the respective min or do you have an alternative naming idea?

FelixKrayer replied:
I myself was very confused when I found out about "min float" being the smallest value greater than zero. Intuitively for me was "min float = - max float" as well.
I think maybe just a different name for it would work? like "neg_max_value" for example.
"`Float_t.neg_max_value`" might be nicer than "`Float_t.sub Nearest Float_t.zero Float_t.max_value`" ?
But I am not sore about the best action here either...

Dudeldu replied:
The problem with `Float_t.neg_max_value` , at least how i see it, is that strictly speaking it had to be defined as `- Float.float_min`. What about `Float_t.upper_bound` and `Float_t.lower_bound` ?

FelixKrayer replied:
Upper bound/lower bound sounds fitting and has no conflict with anything I think

brgr commented on src/cdomains/floatDomain.ml:
```diff
@@ -36,13 +36,17 @@ module type FloatDomainBase = sig
   include Printable.Std (* for default invariant, tag and relift *)
   type t = (float * float) option [@@deriving eq, ord, to_yojson]
 
+  module F32 = CFloat
```
I don't really get this. `F32` is never used. Is the idea here simply that one can replace F32 by F64 if needed?

Dudeldu committed:
Add tests

FelixKrayer commented:
Could you please also include a `min_value` function (You can of course choose a different name for it)? 
`min_value` is actually the least Float, that is **not a subnormal** and I will need it for implementing the `__builtin_isnormal` function (to check if the interval may contain subnormals).
Alternatively `classify_float` should also work, but `min_value` (or similar) would be nicer for me

Dudeldu commented:
This should now be covered by `smallest`, i guess. But the setup also allows extending the interface and adding new c methods easily. Most of the `__builtin_***` could then also be outsourced to "real" c and do not require reassembling them inside ocaml

Dudeldu committed:
Allow using DBL_MIN, DBL_MAX

Some compilers specifiy DBL_MIN, DBL_MAX as long double.
As we do not support long double in neither cil nor goblint itself,
 but want to use DBL_MIN, DBL_MAX we have no other option than to explicitly check for this one case

FelixKrayer commented on src/cdomains/valueDomain.ml:
```diff
@@ -148,7 +148,8 @@ struct
     | TInt (ik,_) -> `Int (ID.top_of ik)
-    | TFloat (FDouble, _) -> `Float (FD.top ())
+    | TFloat (FFloat as fkind, _)
+    | TFloat (FDouble as fkind, _) -> `Float (FD.top_of fkind)
```
I think you could do 
`| TFloat ( (FFloat | FDouble) as fkind , _) -> 'Float (FD.top_of fkind)`
But you decide if that is even an improvement
(If you do, of course do it for the other cases below as well: line 170, 201, 380)

Dudeldu replied:
good idea, that definitely looks better

FelixKrayer reviewed (dismissed):
Looks good

Dudeldu committed:
Improve match patterns

FelixKrayer reviewed (dismissed):
⊤

brgr reviewed (approved):
Please for the next time smaller more separated PRs would be really appreaciated ^^ But looks really good!
